File: Problem Solving/Arrays/2D Array - DS.py
#!/bin/python3

table = [[*map(int, input().split())] for y in range(6)]

def i_vals(y, x):
    yield table[y+1][x+1]
    for x1 in range(3):
        yield table[y+2][x+x1]
        yield table[y][x+x1]

print(max(sum(i_vals(y,x)) for y in range(4) for x in range(4)))

File: Problem Solving/Arrays/2D Array - DS.hs
module Main where

import Data.List.Split
import Control.Monad

main :: IO()
main = replicateM 6 parse >>= print.solve

parse :: IO [Integer]
parse = map read.splitOn " " <$> getLine

solve :: [[Integer]] -> Integer
solve t = maximum [sum $ vals y x | y <- sets, x <- sets]
  where vals = i_vals t; sets = [0..3]

i_vals :: Num a => [[a]] -> Int -> Int -> [a]
i_vals t y x = t!!(y+1)!!(x+1) : [t!!y1!!x1 |
  x1 <- [x..x+2], y1 <- [y,y+2]]
